### PR TITLE
[doc] explaining the integration with huggingface

### PR DESCRIPTION
**Pull request by youkaichao**

### CODE REVIEW

File: docs/source/design/huggingface_transformers.rst [hunk 1]
```diff
@@ -0,0 +1,8 @@
+Integration with 🤗 Transformers
+===================================
+
+This document describes how vLLM integrates with 🤗 Transformers. We will explain step by step what is happening under the hood when we run `vllm serve`.
+
+Let's say we want to serve the popular llama model from 🤗 Transformers by ``vllm serve meta-llama/Llama-3.1-8B```:
```
Comment (russellb) [inline docs/source/design/huggingface_transformers.rst:6]:
```suggestion
Let's say we want to serve the popular llama model from 🤗 Transformers by running ``vllm serve meta-llama/Llama-3.1-8B```:
```

**Review by russellb — COMMENTED** [review]
I know you just started and this is a draft. I noticed one small thing while reading what was here so far.

File: docs/source/design/huggingface_transformers.rst (hunk 1 above)
Comment (youkaichao) [reply docs/source/design/huggingface_transformers.rst]: thanks!

**Comment by youkaichao** [comment]
this doc is mainly the outcome of investigation by @CRZbulabula

**Review by DarkLight1337 — APPROVED** [review]
Thanks for the detailed explanation!

File: docs/source/design/huggingface_integration.rst [hunk 2]
```diff
@@ -0,0 +1,39 @@
+Integration with HuggingFace
+===================================
+
+This document describes how vLLM integrates with HuggingFace libraries. We will explain step by step what happens under the hood when we run ``vllm serve``.
+
+Let's say we want to serve the popular llama model by running ``vllm serve meta-llama/Llama-3.1-8B``.
+
+- The ``model`` argument is ``meta-llama/Llama-3.1-8B``. vLLM will first try to locate the config file ``config.json`` using this argument. See the `code snippet <https://github.com/vllm-project/vllm/blob/10b67d865d92e376956345becafc249d4c3c0ab7/vllm/transformers_utils/config.py#L75>`__ for the implementation.
```
Comment (ywang96) [inline docs/source/design/huggingface_integration.rst:8]: Should we consider making the top level a numbered/ordered list?

File: docs/source/design/huggingface_integration.rst [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+   - If the ``model`` argument is a local path, vLLM will directly read the config file from the path.
+
+   - Otherwise, vLLM will try to read the config from the HuggingFace cache. See `their website <https://huggingface.co/docs/huggingface_hub/en/package_reference/environment_variables#hfhome>`__ for more information on how the HuggingFace cache works. Here, we can also use the argument ``--revision`` to specify the revision of the model in the cache.
+
+   - If neither of the above works, vLLM will download the config file from the HuggingFace model hub, using the ``model`` argument as the model name, the ``--revision`` argument as the revision, and the environment variable ``HF_TOKEN`` as the token to access the model hub. In our case, vLLM will download the ``config.json`` file from ``https://huggingface.co/meta-llama/Llama-3.1-8B/blob/main/config.json``.
```
Comment (ywang96) [inline docs/source/design/huggingface_integration.rst:14]: Perhaps let's not use a gated model for this example? This link will give 404 if the user is not logged in with permission to this model.

File: docs/source/design/huggingface_integration.rst [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+- Model weight: vLLM downloads the model weight from the HuggingFace model hub using the ``model`` argument as the model name and the ``revision`` argument as the revision. vLLM provides the argument ``--load-format`` to control what files to download from the model hub. By default, it will try to load the weights in the safetensors format and fall back to the PyTorch bin format if the safetensors format is not available. We can also pass ``--load-format dummy`` to skip downloading the weights.
```
Comment (ywang96) [inline docs/source/design/huggingface_integration.rst:34]: I would make both `Tokenizer` and `Model weight` bold.

File: docs/source/design/huggingface_integration.rst [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+   - It is recommended to use the safetensors format, as it is efficient for loading in distributed inference and also safe from arbitrary code execution. See the `documentation <https://huggingface.co/docs/safetensors/en/index>`__ for more information on the safetensors format.
```
Comment (ywang96) [inline docs/source/design/huggingface_integration.rst:35]: You need a blank line before this - otherwise the whole parent bullet will be bold.

**Review by ywang96 — COMMENTED** [review]
This is great and we have definitely been asked a lot about how model configuration works on vLLM. I left some comments and PTAL!

File: docs/source/design/huggingface_integration.rst (hunk 2 above)
Comment (youkaichao) [reply docs/source/design/huggingface_integration.rst]: good point, added in [9d81d85](https://github.com/vllm-project/vllm/pull/10173/commits/9d81d852710006ed462494728b7ed0f6826b5105)

File: docs/source/design/huggingface_integration.rst (hunk 3 above)
Comment (youkaichao) [reply docs/source/design/huggingface_integration.rst]: good point, changed to [05f3374](https://github.com/vllm-project/vllm/pull/10173/commits/05f3374c021a998ace5fd53580f4d4b534a09bf0)

File: docs/source/design/huggingface_integration.rst (hunk 4 above)
Comment (youkaichao) [reply docs/source/design/huggingface_integration.rst]: added in [2de7a78](https://github.com/vllm-project/vllm/pull/10173/commits/2de7a78c4638f1010ec83c0e72ae2bee647864ad)

File: docs/source/design/huggingface_integration.rst (hunk 5 above)
Comment (youkaichao) [reply docs/source/design/huggingface_integration.rst:36]: nice catch!

**Comment by youkaichao** [comment]
@ywang96 thanks for the great review and suggestions!

**Comment by russellb** [comment]
Great job on this doc!

**Pull request merged**